lib/runner/concurrency/worker-task.js

```diff
@@ -85,7 +85,10 @@ class WorkerTask extends EventEmitter {
     };
     port2.unref();
 
-    return this.piscina.run({argv: this.argv, port1}, {transferList: [port1]})
+    // remove all the functions from settings, otherwise piscina will fail.
+    const settings = JSON.parse(JSON.stringify(this.settings));
```

Comment:
We don't need to send settings this way as it will cause trouble where we actually do require functions. Instead we could `SafeJson` from utils while exporting results using `postMessage`.

Reply:
We cannot use `SafeJson` here as the `postMessage` is called from `piscina` package and not from Nightwatch.

Also, it shouldn't cause any problem for the test runs on the users' end, as using `JSON.parse(JSON.stringify(this.settings))` entirely removes the properties containing functions, and so those properties will reappear when the settings object is made for each worker (as settings are created using `lodash.merge`).

The only trouble it will have is with the custom settings we pass to `Nightwatch.runTests()` in our own tests when testing parallel testing workflow. If those custom settings contain a function, it will be removed here and will never reappear again in the workers and hence those settings will be lost.

I tried but could not find a way to send functions from the main process to workers, so this is the solution I currently have. Please let me know if there's any other approach you think we can take here.

Reply:
It is called from Nightwatch you can find it here: https://github.com/nightwatchjs/nightwatch/blob/e7fde68bdf6dd43e9a1c7bcd38263f54dd287bda/lib/testsuite/index.js#L670. And we should use safeJSon.  `Json.stringify` doesn't handle circular objects.

Reply:
Actually this `postMessage` is used to send reports from workers to the main process (after the test has finished), and not to send initial data from the main process to workers (while creating a new worker). That `postMessage` exists in the `piscina` package here:

https://github.com/piscinajs/piscina/blob/b0652b4dad9a982957ea993a10fff76e38638abd/src/index.ts#L465

I still tried to use `SafeJSON.stringify()` here and then running a few e2e tests in parallel mode, and while the tests seem to run fine, the HTML reporter shows this error (which goes away on removing the `SafeJSON.stringify` from here):

![image](https://github.com/nightwatchjs/nightwatch/assets/39924567/19c15a7b-40de-4e17-8ad0-5a920e6269d5)

Reply:
What if we use `JSON.parse(SafeJSON.stringify(this.settings))` instead of `JSON.parse(JSON.stringify(this.settings))`? Would that work?

(I tried this and the tests are running fine.)

Reply:
The HTML reporter didn't work as we needed to parse the JSON again upon receiving it on the main thread. You can find the code for that here: https://github.com/nightwatchjs/nightwatch/blob/e7fde68bdf6dd43e9a1c7bcd38263f54dd287bda/lib/reporter/global-reporter.js#L84

Reply:
Okay, I understand what you mean now, and it seems a better approach as it will ensure that `postMessage` never throws an error due to the presence of a function.

But regarding your first comment in this thread:
> We don't need to send settings this way as it will cause trouble where we actually do require functions.

This will still be true because we would still need to do `JSON.parse` which omits all the properties having a function definition as value.

Reply:
Yes, we don't need to send any settings to workers when running parallel e2e tests, because there are no custom settings involved and all required settings can be parsed from the Nightwatch configuration file again by workers.

But while testing the parallel testing workflow in our unit tests, we send some custom settings (like location of custom commands, webdriver settings, etc.) to be used during the test run, and these custom settings if not passed down to workers while creating them, would not appear in workers at all.

Reply:
That worked!